src/common/utxobased/db/makeNewBaselets.ts

```diff
@@ -0,0 +1,189 @@
+  tx: <E extends Executor<'tx'>>(fn: E) => Promise<ReturnType<E>>
+  utxo: <E extends Executor<'utxo'>>(fn: E) => Promise<ReturnType<E>>
+  block: <E extends Executor<'block'>>(fn: E) => Promise<ReturnType<E>>
+
+  all: AddressTables & TransactionTables & UTXOTables
+}
+
+type Executor<DatabaseName extends keyof Databases> = (
+  tables: Databases[DatabaseName]
+) => Promise<unknown>
+
+export interface Databases {
+  address: AddressTables
+  tx: TransactionTables
+  utxo: UTXOTables
+  block: BlockTables
+}
+
+export interface AddressTables {
+  addressByScriptPubkey: HashBase
+  scriptPubkeyByPath: CountBase
+  lastUsedByFormatPath: HashBase
+}
+
+export interface TransactionTables {
+  txById: HashBase
+  txsByScriptPubkey: HashBase
+  txIdsByBlockHeight: RangeBase
+  txsByDate: RangeBase
+}
+
+export interface UTXOTables {
+  utxoById: HashBase
+}
+
+export interface BlockTables {
+  blockHashByBlockHeight: HashBase
+}
+
+export const makeNewBaselets = async (
+  config: MakeBaseletsConfig
+): Promise<Baselets> => {
+  const countBases = await Promise.all([
+    createOrOpen(config.disklet, scriptPubkeyByPathConfig)
+  ])
+  const rangeBases = await Promise.all([
+    createOrOpen(config.disklet, txIdsByBlockHeightConfig),
+    createOrOpen(config.disklet, txsByDateConfig)
+  ])
+  const hashBases = await Promise.all([
+    createOrOpen(config.disklet, addressByScriptPubkeyConfig),
+    createOrOpen(config.disklet, txByIdConfig),
+    createOrOpen(config.disklet, txsByScriptPubkeyConfig),
+    createOrOpen(config.disklet, utxoByIdConfig),
+    createOrOpen(config.disklet, blockHashByBlockHeightConfig),
+    createOrOpen(config.disklet, lastUsedByFormatPathConfig)
+  ])
+
+  const [scriptPubkeyByPath] = countBases
+  const [txIdsByBlockHeight, txsByDate] = rangeBases
+  const [
+    addressByScriptPubkey,
+    txById,
+    txsByScriptPubkey,
+    utxoById,
+    blockHashByBlockHeight,
+    lastUsedByFormatPath
+  ] = hashBases
+
+  const addressBases: AddressTables = {
+    addressByScriptPubkey,
+    scriptPubkeyByPath,
+    lastUsedByFormatPath
+  }
+
+  const txBases: TransactionTables = {
+    txById,
+    txsByScriptPubkey,
+    txIdsByBlockHeight,
+    txsByDate
+  }
+
+  const utxoBases: UTXOTables = {
+    utxoById
+  }
+
+  const blockHeightBases: BlockTables = {
+    blockHashByBlockHeight
+  }
+
+  return {
+    async address<E extends Executor<'address'>>(
+      fn: E
+    ): Promise<ReturnType<E>> {
+      return await fn(addressBases)
+    },
+
+    async tx<E extends Executor<'tx'>>(fn: E): Promise<ReturnType<E>> {
+      return await fn(txBases)
+    },
+
+    async utxo<E extends Executor<'utxo'>>(fn: E): Promise<ReturnType<E>> {
+      return await fn(utxoBases)
+    },
+
+    async block<E extends Executor<'block'>>(fn: E): Promise<ReturnType<E>> {
+      return await fn(blockHeightBases)
+    },
+
+    all: {
+      ...addressBases,
+      ...txBases,
+      ...utxoBases,
+      ...blockHeightBases
+    }
+  }
```

Comment:
This was badly broken in my editor, and then I realized we really shouldn't be duplicated the type information within the implementation (`Baselets` should contain all the type information we need to type this return value). However, TypeScript wont be happy without any type annotations per method, the minimum I was able to achieve is:

```ts
return {
    async address(fn): Promise<ReturnType<typeof fn>> {
      return await fn(addressBases)
    },

    async tx(fn): Promise<ReturnType<typeof fn>> {
      return await fn(txBases)
    },

    async utxo(fn): Promise<ReturnType<typeof fn>> {
      return await fn(utxoBases)
    },

    async block(fn): Promise<ReturnType<typeof fn>> {
      return await fn(blockHeightBases)
    },

    all: {
      ...addressBases,
      ...txBases,
      ...utxoBases,
      ...blockHeightBases
    }
  }
```

Reply:
Used what you suggested here.